Improve error handling by introducing exceptions.

kasperg opened the pull request:
Currently the Apsis class does not allow the caller to distinguish
Between different types of errors which may occur in the communication
with the APSIS API. All exceptions are caught and FALSE returned.

This change introduces a mapper which handles the rather complex logic
Of mapping from generic HTTP request exceptions to specific APSIS
Exceptions based on custom APSIS error codes and messages as well as 
HTTP status codes.

API callers are updated to handle the new exceptions and act 
accordingly.

Most importantly this means that we can handle queued subscriptions
correctly. If an email is on an opt-out list or the email is invalid
then there is no need to try to process the subscription at a later
point in time. Thus we avoid blocking the queue.

#SAMD8-578

xendk commented on src/Apsis.php:
```diff
@@ -90,20 +110,18 @@ protected function request($method, $path, array $args = []) {
       // Try request.
       try {
         // Do http request.
+        /* @var \Psr\Http\Message\ResponseInterface$response */
```
Missing space?

kasperg replied:
:+1:

xendk commented on src/Exception/BadRequestException.php:
```diff
@@ -0,0 +1,30 @@
+ * Created by PhpStorm.
+ * User: kasper
+ * Date: 10/01/2018
+ * Time: 14.38
```
Isn't these kinda useless? Doesn't follow the coding standard either.

kasperg replied:
Totally agree. I missed that. I blame it on PhpStorm + comments folding.

xendk commented on src/Exception/ExceptionMapper.php:
```diff
@@ -0,0 +1,180 @@
+/**
+ * Mapper for transforming generic HTTP exceptions to specific ApsisExceptions.
+ *
+ * Officially APSIS uses a number of different types of errors. This set is larger than the number of used HTTP status
```
"In general, all lines of code should not be longer than 80 characters."

kasperg replied:
:+1:

xendk commented on src/Plugin/QueueWorker/AddSubscriber.php:
```diff
@@ -53,19 +68,20 @@ public static function create(ContainerInterface $container, array $configuratio
+      // Invalid data. Nothing really we can do here. Log and more on.
+      $this->logger->error(sprintf('Unable to subscribe email %s to list %s: %s', $email, $list_id, $e->getMessage()));
+    } catch (ApiDisabledException $e) {
+      throw new SuspendQueueException($e);
```
:+1:

xendk commented:
Like the idea of a mapper, but I think the implementation is a bit too generic. It could be more readable with some simple loops rather than array_reduce and static getters rather than instantiating all exceptions twice.

kasperg commented:
@tripox: I removed the unneeded comments.

@xendk I updated the mapping code. It no longer creates exception instances multiple times to compare. Instead we use static getters. This also aligns with the approach for getting the match phrase. I also replaced array_reduce with a simpler loop.

I also added a unit test 😱
